lib/publiccloud/provider.pm

```diff
@@ -705,6 +705,25 @@ sub terraform_param_tags
     return encode_json($tags);
 }
 
+=head2 get_terraform_output
+
+Query the terraform data structure in json format.
+Input: <jq-query-format> string; <empty> = no query then full output of data structure.
+E.g: to get the VM instance name from json data structure, the call is: 
+    get_terraform_output(".vm_name.value[0]");
+To get the complete output structure, the call is:
+    get_terraform_output();
+
+=cut
+
+sub get_terraform_output {
+    my ($self, $jq_query) = @_;
+    assert_script_run('cd ~/terraform');
+    my $res = script_output("terraform output -json | jq -r \"$jq_query\" 2>/dev/null", proceed_on_failure => 1);
```

Comment:
I would use single-quotes here because JSON queries can use double quotes...

```suggestion
    my $res = script_output("terraform output -json | jq -r '$jq_query' 2>/dev/null", proceed_on_failure => 1);
```

Reply:
ok, but my concern was that single quotes prevented the inner variable expansion

Reply:
+ Single and double quotes have different meanings in bash ( same as Perl fun )

Reply:
yes, but at that level the perl syntax comes first, so question: will that perl variable be safe with single quotes?

Reply:
if they are inside double-quoted string they should be taken as normal characters so Perl var should correctly eval

```Perl
$ use Modern::Perl
$ my $test = "testuji string"
testuji string
$ say "test - '$test'"
test - 'testuji string'
1
$ say 'test - "$test"'
test - "$test"
```

Reply:
ok :smile:

Reply:
But I have different results in my oneliner:
```
perl -e '
my $q = "This is a test STRING.";
print "Testing: \"$q\"";
print "\n";
'
```
`Testing: "This is a test STRING."`
```
perl -e '
my $q = "This is a test STRING.";
print "Testing: '$q'";
print "\n";
'
```
`Testing:`

but even failure with:
```
perl -e '
my $q = "This is a test STRING.";
print 'Testing: "$q" ';
print "\n";
'
```
`syntax error at -e line 3, near "Testing:"
Execution of -e aborted due to compilation errors.`

Therefore I should run on OSD that change porposal, if really it works.

Reply:
@m-dati that's because the `'` is ending your perl input and then the shell tries to resolve the environment variable `$q`.
Try this:
```
perl -e '
my $q = "This is a test STRING.";
print "Testing: '$USER'";
print "\n";
'
```
Also, as a general note, always use `-w` in oneliners. I have seen so many questions over the years where the answer was simple "enable warnings and then you see yourself why it doesn't work". (not in this case though)
Also, use `-E` in oneliners, so you can just use `say` and save the `"\n"` thing.
e.g.
```
perl -wE '
my $q = "This is a test STRING.";
say "Testing: '$USER'";
'
```

Reply:
Ok, in the editor I have:
```
...
print "Testing1: ==>'$q' --> $q ";
```
`Testing1: ==>'This is a test STRING.' --> This is a test STRING.`

Updating.

Reply:
@perlpunk sorry I late reading your reply: yes as I wrote, I noted the external quote impact :slightly_smiling_face:. But indeed your suggestions on options are really useful. Thank you!